src/libos/crates/async-rt/src/time/wheel.rs

```diff
@@ -0,0 +1,103 @@
+use super::entry::TimerWheelEntry;
+use super::Instant;
+use crate::prelude::*;
+use spin::mutex::MutexGuard;
+
+use hierarchical_hash_wheel_timer::wheels::quad_wheel::QuadWheelWithOverflow;
+
+lazy_static! {
+    pub static ref TIMER_WHEEL: TimerWheel = {
+        use crate::executor::EXECUTOR;
+        use crate::sched::yield_;
+        use crate::task::Task;
+
+        let wheel = TimerWheel::new();
+
+        // TODO: Don't always run this task when no useful tasks to run and no timeouts to expire.
+        let task = Arc::new(Task::new(async {
+            loop {
+                TIMER_WHEEL.try_make_progress();
+                yield_().await;
+            }
+        }));
+        EXECUTOR.accept_task(task);
+
+        wheel
+    };
+}
+
+/// The interface of hierarchical timer wheel. The resolution of time is 1ms.
+pub struct TimerWheel {
+    // hierarchical timer wheel.
+    wheel: Mutex<QuadWheelWithOverflow<TimerWheelEntry>>,
+    // current ticks, one tick means 1ms.
+    ticks: AtomicU64,
+    // start time of the timerwheel.
+    start: Instant,
+}
+
+impl TimerWheel {
+    pub fn new() -> Self {
+        Self {
+            wheel: Mutex::new(QuadWheelWithOverflow::default()),
+            ticks: AtomicU64::new(0),
+            start: Instant::now(),
+        }
+    }
+
+    /// Insert a new timer entry to the wheel.
+    /// Return the latest tick of the timerwheel, or Err() if the duration is too short (less than 1ms).
```

Comment:
Do you think it is a good idea to round up the timeout of `entry` to 1ms so that the insertion can always succeed? Instead of documenting about all possible errors of an API, it is more preferable to eliminate errors in the first place. Infallible APIs are most user friendly. We can even guarantee this property of minimal timeout value by changing the constructor of `TimerEntry`.

By the way, it seems that you have checked the property in a later phase.

```rust
                // The minimum resolution of QuadWheelWithOverflow is 1 ms.
                // If the timeout less than 1 ms, wait 1ms instead.
                let timeout = core::cmp::max(shared.timeout, Duration::MILLISECOND);
```

Reply:
Since I checked the property before insertion, the insertion will always succeed.
I agree to move the check to insertion function, hence the insertion will always succeed and doesn't need the comments anymore.

Reply:
```Rust
    /// Insert a new timer entry to the wheel and return the ticks when the entry is inserted.
    pub fn insert_entry(&self, entry: TimerWheelEntry, timeout: Duration) -> u64 {
        let mut guard = self.wheel.lock();

        // Try to make progress to assure that the wheel is up to date.
        let entries = self.make_progress(&mut guard);

        // The minimum resolution of QuadWheelWithOverflow is 1 ms.
        // If the timeout less than 1 ms, wait 1ms instead.
        let timeout = core::cmp::max(timeout, Duration::MILLISECOND);
        guard.insert_with_delay(entry, timeout).unwrap();
        let insert_ticks = self.latest_ticks();

        drop(guard);
        Self::fire(entries);

        insert_ticks
    }
```

Reply:
Just to be sure. The timeout value being rounded up is not visible to users, right? For example, the return value of `remained_duration ` method.